src/main/java/appeng/recipes/handlers/InscriberProcess.java

```diff
@@ -0,0 +1,115 @@
+package appeng.recipes.handlers;
+
+
+import java.util.List;
+import javax.annotation.Nullable;
+
+import net.minecraft.item.ItemStack;
+
+import appeng.api.exceptions.MissingIngredientError;
+import appeng.api.exceptions.RecipeError;
+import appeng.api.exceptions.RegistrationError;
+import appeng.api.recipes.ICraftHandler;
+import appeng.api.recipes.IIngredient;
+import appeng.recipes.RecipeHandler;
+import appeng.util.Platform;
+
+
+/**
+ * basic inscriber process for recipes
+ *
+ * @author AlgorithmX2
+ * @author thatsIch
+ * @version rv2
+ * @since rv0
+ */
+public abstract class InscriberProcess implements ICraftHandler, IWebsiteSerializer
+{
+	@Nullable
+	private IIngredient imprintable;
+
+	@Nullable
+	private IIngredient topOptional;
+
+	@Nullable
+	private IIngredient botOptional;
+
+	@Nullable
+	private IIngredient output;
+
+	@Override
+	public void setup( List<List<IIngredient>> input, List<List<IIngredient>> output ) throws RecipeError
+	{
+		if( output.size() == 1 && output.get( 0 ).size() == 1 )
+		{
+			if( input.size() == 1 && input.get( 0 ).size() > 1 )
+			{
+				this.imprintable = input.get( 0 ).get( 0 );
+
+				this.topOptional = input.get( 0 ).get( 1 );
+
+				if( input.get( 0 ).size() > 2 )
+					this.botOptional = input.get( 0 ).get( 2 );
+
+				this.output = output.get( 0 ).get( 0 );
+			}
+			else
+				throw new RecipeError( "Inscriber recipes cannot have rows, and must have more then one input." );
+		}
+		else
+			throw new RecipeError( "Inscriber recipes must produce a single output." );
+	}
+
+	@Override
+	public boolean canCraft( ItemStack reqOutput ) throws RegistrationError, MissingIngredientError
```

Comment:
if the person entered a recipe wrongly

like 

    inscribe stuff -> output

missing an optional